### PR TITLE
Test Arraylist Implementation

### PR DESCRIPTION
**Pull request by valeriuchicu**
Test ArrayList Implementation

### CODE REVIEW

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class StudentListTest {
+
+    @BeforeEach
+    public void setUp() {
+    }
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: what does this method do ?

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 2]
```diff
@@ -0,0 +1,870 @@
...
+
+    @Test
+    void constructorShouldThrowAnExceptionIfTheCapacityIsLessThenZero() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: please name all tests by the following convention
shouldExpectedBehaviorWhenStateUnderTest
ex: 
shouldThorowIllegalArgumentExceptionWhenCreatingNewInstanceWithCapacityLessThenZero

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 3]
```diff
@@ -0,0 +1,870 @@
...
+
+        //when
+        int listSize = studentList.size();
+
+        //then
+        assertEquals(listSize, 3);
+    }
+
+    @Test
+    void isArrayEmpty() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: ex:    shouldReturnFalseWhenTheListHasNoElements

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 4]
```diff
@@ -0,0 +1,870 @@
...
+
+        //given
+        List<Student> studentList = new StudentList<>();
+        //when
+        //then
+        assertTrue(studentList.isEmpty());
+    }
+
+    @Test
+    void arrayShouldBeNotEmpty() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: ex:   shouldReturnTrueWhenTheListHasElements

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+        int listSize = studentList.size();
+
+        //then
+        assertEquals(listSize, 3);
+    }
+
+    @Test
+    void isArrayEmpty() {
+
+        //given
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: please remove all the comments

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+        studentList.add(st1);
+        studentList.add(st2);
+        studentList.add(st3);
+
+        list.add(st1);
+        list.add(st2);
+        list.add(st3);
+
+        //when
+        Object[] actualArrayOfStudents = studentList.toArray();
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java:122]: Initialize actualArrayOfStudents directly with an array, do not use ArrayList

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 7]
```diff
@@ -0,0 +1,870 @@
...
+        boolean existSt1 = studentList.contains(st1);
+        boolean existSt2 = studentList.contains(st2);
+
+        //then
+        assertTrue(existSt1);
+        assertTrue(existSt2);
+    }
+
+    @Test
+    void arrayShouldNotContainsElement() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: this is not an array

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 8]
```diff
@@ -0,0 +1,870 @@
...
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+        studentList.add(st1);
+        studentList.add(st2);
+        studentList.add(st3);
+
+        //when
+        Student removedStudent = studentList.remove(1);
+        boolean isContainTheStudent = studentList.contains(removedStudent);
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: you could assert this condition in the test above, it's related to delete.
No need for two separate tests here

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 9]
```diff
@@ -0,0 +1,870 @@
...
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+        studentList.add(st1);
+        studentList.add(st2);
+        studentList.add(st3);
+
+        //when
+        //then
+        assertThrows(IndexOutOfBoundsException.class, () -> studentList.remove(-1));
+        assertThrows(IndexOutOfBoundsException.class, () -> studentList.remove(3));
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: assert the thrown message, as well

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 10]
```diff
@@ -0,0 +1,870 @@
...
+    void ifInsertedIndexInMethodGetIsOutOfBound() {
+        //given
+        List<Student> studentList = new StudentList<>();
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        studentList.add(st1);
+        studentList.add(st2);
+        //when
+        //then
+        assertThrows(IndexOutOfBoundsException.class, () -> studentList.get(3));
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java:176]: assert the thrown message

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 11]
```diff
@@ -0,0 +1,870 @@
...
+        List<Student> studentList = new StudentList<>();
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+        studentList.add(st1);
+        studentList.add(st2);
+        //when
+        Student setStudent = studentList.set(1, st3);
+        //then
+        assertEquals(st3, setStudent);
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: you do not assert what the name says, how do you know that it has been added at the correct index ?
is st2 still in the list ?

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 12]
```diff
@@ -0,0 +1,870 @@
...
+        List<Student> studentList = new StudentList<>();
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+        studentList.add(st1);
+        studentList.add(st2);
+        //when
+        studentList.add(1, st3);
+        //then
+        assertEquals(st3, studentList.get(1));
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java:204]: same as the one above

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 13]
```diff
@@ -0,0 +1,870 @@
...
+            List<Student> studentList = new StudentList<>();
+            Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+            Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+            Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+            studentList.add(st1);
+            studentList.add(st2);
+            studentList.add(st3);
+            ListIterator<Student> listIterator = studentList.listIterator();
+            //when
+//            listIterator.next();
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: ?

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 14]
```diff
@@ -0,0 +1,870 @@
...
+            assertEquals(st1, studentList.get(1));
+            assertEquals(st2, studentList.get(2));
+            assertEquals(st3, studentList.get(3));
+        }
+
+        @Test
+        void addShouldAddTheElementAtTheIndexTree() {
+            //given
+            List<Student> studentList = new StudentList<>();
+//            List<Student> studentList = new ArrayList<>();
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: remove all comments or commented out code

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 15]
```diff
@@ -0,0 +1,870 @@
...
+    @Test
+    void shouldTheSizeBeRight() {
+        //given
+        List<Student> studentList = new StudentList<>();
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        Student st3 = new Student("Trei", LocalDate.of(2006, 1, 1), "Student TREE");
+        studentList.add(st1);
+        studentList.add(st2);
+        studentList.add(st3);
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java:33]: duplicate code.
many of your tests use the same or similar data.
extract the studentList  at instance level
make use of @BeforeEach @AfterEach, private utility methods to populate test data

**Comment by raileanserj** [comment]
Please add at least one Parameterized Test

**Comment by raileanserj** [comment]
Try to improve StudentList.java line coverage, current is 94%, which is not bad, however it's not as good as 100% )

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 1 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: You are right, I will use this method to populate the test data. 
Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 3 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 4 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 15 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:33]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 1 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:22]: After I saw your last suggestion I used this method to populate the test data.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 14 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 5 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 12 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:204]: I am sorry, but I cannot find another method with the same logic.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 13 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: I just test with what happened if  I call listIterator.next();
I deleted it. 
Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 11 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: You are right. I corrected it. I set the object at the index 1 and after that I get the object also from the index 1. 
Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 2 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 6 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:122]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 9 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 10 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:176]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 7 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 8 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+        Student st1 = new Student("Unu", LocalDate.of(2001, 1, 1), "Student ONE");
+        Student st4 = new Student("Patru", LocalDate.of(2006, 1, 1), "Student PATRU");
+        return Stream.of(
+                Arguments.of(st1, true),
+                Arguments.of(st4, false),
+                Arguments.of(null, false));
+    }
+
+    @Nested
+    class testAllMethodsOfTheInnerClassIterator {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: class name should start with capital letter

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+
+    @Test
+    void shouldCreateAListWithDefaultCapacity() {
+        List<Student> studentList = new StudentList<>(8);
+
+        assertNotNull(studentList);
+    }
+
+    @Test
+    void shouldCreatListWithLengthFifteen() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: the list does not have length, it has size, and it is not 15.
better remove this test

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 18]
```diff
@@ -0,0 +1,551 @@
...
+    }
+
+    @Test
+    void shouldThrowIllegalArgumentExceptionWithInvalidParameter() {
+        Throwable exceptionThatWasThrown = assertThrows(IllegalArgumentException.class, () -> new StudentList<Student>(-1));
+        assertEquals(exceptionThatWasThrown.getMessage(), "Illegal Capacity: " + "-1");
+    }
+
+    @Test
+    void shouldCreateAListWithDefaultCapacity() {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: you cannot prove that the list has been create with default capacity. 
Better remove this test

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 16 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done. Thank you.

File: src/test/java/com/endava/internship/collections/StudentListTest.java [hunk 19]
```diff
@@ -0,0 +1,551 @@
...
+    @Test
+    void shouldStartFromTheGivenIndex() {
+        Student st2 = new Student("Doi", LocalDate.of(2004, 1, 1), "Student TWO");
+        boolean theListIteratorStartFromIndex = st2.equals(studentList.listIterator(1).next());
+
+        assertTrue(theListIteratorStartFromIndex);
+    }
+
+    @Nested
+    class testAllMethodsOfTheInnerClassListIterator {
```
Comment (raileanserj) [inline src/test/java/com/endava/internship/collections/StudentListTest.java]: class name should start with capital letter

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 17 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 18 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 19 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java]: Done.

File: src/test/java/com/endava/internship/collections/StudentListTest.java (hunk 18 above)
Comment (valeriuchicu) [reply src/test/java/com/endava/internship/collections/StudentListTest.java:45]: Done.

**Pull request merged**